modules/navier_stokes/src/fvkernels/INSFVTKEDSourceSink.C

```diff
@@ -0,0 +1,253 @@
+            (velocity - velocity * _normal[_current_elem][i] * _normal[_current_elem][i]).norm();
+
+        y_plus = NS::findyPlus(_mu(makeElemArg(_current_elem), state),
+                               _rho(makeElemArg(_current_elem), state),
+                               std::max(parallel_speed, 1e-10),
+                               distance);
+      }
+
+      y_plus_vec.push_back(y_plus);
+
+      tot_weight += 1.0;
+    }
+
+    for (unsigned int i = 0; i < y_plus_vec.size(); i++)
+    {
+      auto y_plus = y_plus_vec[i];
+
+      if (y_plus < 11.25)
+      {
+        auto fi = _face_infos[_current_elem][i];
+        const bool defined_on_elem_side = _var.hasFaceSide(*fi, true);
+        const Elem * const loc_elem = defined_on_elem_side ? &fi->elem() : fi->neighborPtr();
+        Moose::FaceArg facearg = {
+            fi, Moose::FV::LimiterType::CentralDifference, false, false, loc_elem};
+        ADReal wall_mut = _mu_t(facearg, state);
+        destruction += 2.0 * _k(makeElemArg(_current_elem), state) * wall_mut /
+                       Utility::pow<2>(_dist[_current_elem][i]) / tot_weight;
+      }
+      else
+        destruction += std::pow(_C_mu, 0.75) * _rho(makeElemArg(_current_elem), state) *
+                       std::pow(_k(makeElemArg(_current_elem), state), 1.5) /
+                       (_von_karman * _dist[_current_elem][i]) / tot_weight;
+    }
+
+    residual = _var(makeElemArg(_current_elem), state) - destruction;
+  }
+  else
+  {
+
+    const auto & grad_u = _u_var->adGradSln(_current_elem, state);
+    auto Sij_xx = 2.0 * grad_u(0);
+    ADReal Sij_xy = 0.0;
+    ADReal Sij_xz = 0.0;
+    ADReal Sij_yy = 0.0;
+    ADReal Sij_yz = 0.0;
+    ADReal Sij_zz = 0.0;
+
+    auto grad_xx = grad_u(0);
+    ADReal grad_xy = 0.0;
+    ADReal grad_xz = 0.0;
+    ADReal grad_yx = 0.0;
+    ADReal grad_yy = 0.0;
+    ADReal grad_yz = 0.0;
+    ADReal grad_zx = 0.0;
+    ADReal grad_zy = 0.0;
+    ADReal grad_zz = 0.0;
+
+    auto trace = Sij_xx / 3.0;
+
+    if (_dim >= 2)
+    {
+      const auto & grad_v = _v_var->adGradSln(_current_elem, state);
+      Sij_xy = grad_u(1) + grad_v(0);
+      Sij_yy = 2.0 * grad_v(1);
+
+      grad_xy = grad_u(1);
+      grad_yx = grad_v(0);
+      grad_yy = grad_v(1);
+
+      trace += Sij_yy / 3.0;
+
+      if (_dim >= 3)
+      {
+        const auto & grad_w = _w_var->adGradSln(_current_elem, state);
+
+        Sij_xz = grad_u(2) + grad_w(0);
+        Sij_yz = grad_v(2) + grad_w(1);
+        Sij_zz = 2.0 * grad_w(2);
+
+        grad_xz = grad_u(2);
+        grad_yz = grad_v(2);
+        grad_zx = grad_w(0);
+        grad_zy = grad_w(1);
+        grad_zz = grad_w(2);
+
+        trace += Sij_zz / 3.0;
+      }
+    }
+
+    auto symmetric_strain_tensor_sq_norm =
+        (Sij_xx - trace) * grad_xx + Sij_xy * grad_xy + Sij_xz * grad_xz + Sij_xy * grad_yx +
+        (Sij_yy - trace) * grad_yy + Sij_yz * grad_yz + Sij_xz * grad_zx + Sij_yz * grad_zy +
+        (Sij_zz - trace) * grad_zz;
```

Comment:
This could be done very easily, if you used tensors/vectors instead of each component separately. 
The tensor in libmesh has this contract function:

https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1TypeTensor.html#aca9325828337e17e35dfad38318f5664

Which will do the multiplication and summation for you. So you have one tensor for (A^T+A - trace*I) and you contract it with A.

Reply:
Is this faster than doing a single component at the time? It was not in my experience...

Reply:
The tensor implementation was my previous one and I had to code one component at the time because it was awfully slow

Reply:
Maybe I am missing something you know that I don't